### PR TITLE
[MM-15425] [MM-13850] Improved plugin development cycle

### PR DESCRIPTION
**Pull request by cpoile**
### Summary
- I'd like to have our team use this during Jira development for a few weeks before making it 'official' in the `plugin-starter-template`, `plugin-demo`, and `plugin-sample` repos.
- Updated with the comments and suggestions from https://github.com/mattermost/mattermost-plugin-starter-template/pull/39
- Webapp watcher now resets the plugin after compiling any changes. Most of the time webapp changes don't require a reload (@asaadmahmood will like this). 
- Webapp watcher runs in the foreground (Jesse's request).
- Everything was moved into the makefile (eg, debug-plugin and kill-plugins)

### Workflow
This is how I've been using these for the last couple months. I will expand this into a tutorial for the docs.
- Export or source your MM env settings (MM_SERVICESETTINGS_SITEURL, MM_ADMIN_USERNAME, MM_ADMIN_PASSWORD)
- `make deploy` like before if you want to reset everything. I usually do this after switching branches, just in case.
- `make sd` (easy-to-type alias for `make server-debug`) to compile the go code (for your architecture only) and upload it to the local server.
- `make wd` (easy-to-type alias for `make webapp-debug`) to set up a webpack watcher. It will upload webapp changes after compiling.
- `make debug-plugin` to attach a headless delve process to the plugin. Point your debugger to localhost:2346
- `make kill-plugin` will kill any plugin processes. Do this when there more than one and the `make debug-plugin` doesn't know which one to attach to.
- To debug the webapp part of the plugin, attach the debugger to localhost:8065 (as you would normally debug the mattermost webapp).

### Links
- Old PR, closed: #70 
- Plugin sample PR with review comments (addressed): https://github.com/mattermost/mattermost-plugin-starter-template/pull/39
- [MM-15425](https://mattermost.atlassian.net/browse/MM-15425)
- server issue: https://github.com/mattermost/mattermost-server/issues/10171
- Solves improvement ticket: [MM-13850](https://mattermost.atlassian.net/browse/MM-13850)

### CODE REVIEW

File: webapp/webpack.config.js [hunk 1]
```diff
@@ -71,4 +80,50 @@ module.exports = {
...
+                });
+            },
+        },
+    ],
 };
+
+// Development mode configuration
+if (DEV) {
+    config.mode = 'development';
+    config.devtool = 'source-map';
```
Comment (crspeller) [inline webapp/webpack.config.js]: Redundant. We always generate source-maps in the config above.

File: Makefile [hunk 2]
```diff
@@ -179,3 +181,117 @@ endif
...
+
+# If we were debugging, we have to unattach the delve process or else we can't disable the plugin.
+# NOTE: we are assuming the dlv was listening on port 2346, as in the debug-plugin.sh script.
+	@DELVE_PID=$(shell ps aux | grep "dlv attach.*2346" | grep -v "grep" | awk -F " " '{print $$2}') && \
+	if [ "$$DELVE_PID" -gt 0 ] > /dev/null 2>&1 ; then \
+		echo "Located existing delve process running with PID: $$DELVE_PID. Killing." ; \
+		kill -9 $$DELVE_PID ; \
+	fi
+
+ifneq ($(CURL),)
```
Comment (crspeller) [inline Makefile]: Maybe have the curl check above with the other precondition checks?

File: Makefile [hunk 3]
```diff
@@ -179,3 +181,117 @@ endif
...
+ifeq ($(OS),Darwin)
+	cd server && env GOOS=darwin GOARCH=amd64 $(GO) build -gcflags "all=-N -l" -o dist/plugin-darwin-amd64;
+else ifeq ($(OS),Linux)
+	cd server && env GOOS=linux GOARCH=amd64 $(GO) build -gcflags "all=-N -l" -o dist/plugin-linux-amd64;
+else ifeq ($(OS),Windows_NT)
+	cd server && env GOOS=windows GOARCH=amd64 $(GO) build -gcflags "all=-N -l" -o dist/plugin-windows-amd64.exe;
+else
+	$(error make debug depends on uname to return your OS. If it does not return 'Darwin' (meaning OSX), 'Linux', or 'Windows_NT' (all recent versions of Windows), you will need to edit the Makefile for your own OS.)
+endif
+
```
Comment (hanzei) [inline Makefile]: I'm a bit concerned that this is basically a copy of `bundle`. Can't you just call `bundle` instead?

File: Makefile [hunk 4]
```diff
@@ -179,3 +181,117 @@ endif
...
+sd: server-debug
+
+# server-debug builds and deploys a debug version of the plugin for your architecture.
+# Then resets the plugin to pick up the changes.
+.PHONY: server-debug
+server-debug: server-debug-deploy reset
+
+.PHONY: server-debug-deploy
+server-debug-deploy:
+
```
Comment (hanzei) [inline Makefile]: Nit: remove this line

File: Makefile [hunk 5]
```diff
@@ -179,3 +181,117 @@ endif
...
+# wd is an easier-to-type alias for webapp-debug
+.PHONY: wd
+wd: webapp-debug
+
+# webapp-debug builds and deploys the plugin's webapp in watch mode with source-maps.
+# Webpack will run make-reset after detecting and compiling changes.
+.PHONY: webapp-debug
+webapp-debug:
+
```
Comment (hanzei) [inline Makefile]: Nit: remove this line

File: Makefile [hunk 6]
```diff
@@ -179,3 +181,117 @@ endif
...
+# NOTE: we are assuming the dlv was listening on port 2346, as in the debug-plugin.sh script.
+	@DELVE_PID=$(shell ps aux | grep "dlv attach.*2346" | grep -v "grep" | awk -F " " '{print $$2}') && \
+	if [ "$$DELVE_PID" -gt 0 ] > /dev/null 2>&1 ; then \
+		echo "Located existing delve process running with PID: $$DELVE_PID. Killing." ; \
+		kill -9 $$DELVE_PID ; \
+	fi
+
+ifneq ($(CURL),)
+	@echo "\nRestarting plugin via API"
+	$(eval TOKEN := $(shell curl -i -X POST $(MM_SERVICESETTINGS_SITEURL)/api/v4/users/login -d '{"login_id": "$(MM_ADMIN_USERNAME)", "password": "$(MM_ADMIN_PASSWORD)"}' | grep -o "Token: [0-9a-z]*" | cut -f2 -d' ' 2> /dev/null))
```
Comment (hanzei) [inline Makefile]: What is the difference between this and `deploy`?

File: Makefile (hunk 3 above)
Comment (cpoile) [reply Makefile]: Sorry, I don't see what you mean -- bundle is pretty different from this. If anything, it's similar to deploy. But here we're only compiling for the user's OS (cutting 2/3 of the time off the build process).

File: Makefile (hunk 6 above)
Comment (cpoile) [reply Makefile]: Between `reset` and `deploy`? I'm not sure what you mean -- they're almost entirely different except that they both upload at the end (and deploy falls-back to copying). For instance, `deploy` calls `dist`, `reset` doesn't.

**Pull request merged**